frontends/PyCDE/src/bsp/xrt.py

```diff
@@ -39,14 +89,85 @@ def XrtBSP(user_module):
   environment), set the PYTHON variable (e.g. 'PYTHON=python3.9').
   """
 
+  class XrtService(Module):
+    clk = Clock(types.i1)
+    rst = Input(types.i1)
+
+    axil_in = Input(axil_in_type(axil_addr_width, axil_data_width))
+    axil_out = Output(axil_out_type(axil_data_width))
+
+    @generator
+    def generate(self):
+      clk = self.clk
+      rst = self.rst
+
+      sys: System = System.current()
+      output_tcl((sys.hw_output_dir / "xrt_package.tcl").open("w"))
+
+      ######
+      # Write side.
+
+      # So that we don't wedge the AXI-lite for writes, just ack all of them.
+      write_happened = Wire(bit)
+      latched_aw = ControlReg(self.clk, self.rst, [self.axil_in.awvalid],
+                              [write_happened])
+      latched_w = ControlReg(self.clk, self.rst, [self.axil_in.wvalid],
+                             [write_happened])
+      write_happened.assign(latched_aw & latched_w)
+
+      ######
+      # Read side.
+
+      # Track the non-zero registers in the read address space.
+      rd_addr_data = {
+          16: Bits(32)(MagicNumberLo),
+          20: Bits(32)(MagicNumberHi),
+          24: Bits(32)(VersionNumber),
+      }
+
+      # Create an array out of the sparse value map 'rd_addr_data' and zero
+      # constants. Then create a potentially giant mux. There's probably a much
+      # better way to do this. I suspect this is a common high-level construct
+      # which should be automatically optimized, but I'm not sure how common
+      # this actually is.
+
+      max_addr_log2 = int(
+          math.ceil(math.log2(max([a for a in rd_addr_data.keys()]) + 1)))
+
+      # Convert the sparse dict into a zero filled array.
+      zero = types.int(axil_data_width)(0)
+      rd_space = [zero] * int(math.pow(2, max_addr_log2))
+      for (addr, val) in rd_addr_data.items():
+        rd_space[addr] = val
+
+      # Create the address index signal and do the muxing.
+      addr_slice = self.axil_in.araddr.slice(
+          types.int(axil_addr_width)(0), max_addr_log2)
+      rd_addr = addr_slice.reg(clk, rst)
+      rvalid = self.axil_in.arvalid.reg(clk, rst)
+      rdata = types.array(types.int(axil_data_width),
+                          len(rd_space))(rd_space)[rd_addr]
```

Comment:
I suggest registering the read data mux output

```suggestion
      rvalid = self.axil_in.arvalid.reg(clk, rst, cycles=2)
      rdata = types.array(types.int(axil_data_width),
                          len(rd_space))(rd_space)[rd_addr].reg(clk)
```